### ISSUE TITLE
Sharing an NVMe SSD with Multiple Processes

### ISSUE THREAD

**Issue by yuezhu1** [spdk/spdk#566]
multiple processes can share an SSD if they have the same shared memory group ID. I noticed an error (out of memory for nvme request submission when io queue depth = 64, and transfer size = 1 MB) reported by the built-in perf tool when running perf with nvme-of on the same node. 

In detail:
After I started NVMe-oF, I allow the built-in tool perf to access the same SSD for performance profiling. If I set the io queue depeth as 64 and transfer size = 1 MB, the 8 starting I/O failures (the last 8 of 64 in the first round of submission) are caused by spdk_nvme_ns_cmd_read_with_md. The error is -ENOMEM. It seems like there is not enough memory for the submission queue. However, the queue length is 256 (obtained by spdk_nvme_ctrlr_get_default_io_qpair_opts()). I wonder if anyone has any ideas about how to solve this error.

## Expected Behavior

I expect the perf can submit the read requests without any failure when io queue depth = 64 and transfer size = 1 MB.

## Current Behavior

the 8 starting I/O failures (the last 8 of 64 in the first round of submission) are caused by spdk_nvme_ns_cmd_read_with_md. 

If I only specify the memory usage size in the primary process, I still get the same problem.

## Steps to Reproduce

Only specify the memory usage size in NVMe-oF

1. The output for nvme-of:
./nvmf_tgt -c nvmf.conf.in -i 1 -s 2048
Starting SPDK v19.01-pre / DPDK 18.08.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 -m 2048 --base-virtaddr=0x200000000000 --file-prefix=spdk1 --proc-type=auto ]
EAL: Detected 40 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: Auto-detected process type: PRIMARY
EAL: Multi-process socket /run/user/1003/dpdk/spdk1/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
app.c: 612:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 298:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
EAL: PCI device 0000:04:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme

2../perf -q 64 -o 1048576 -w read -t 5 -r 'trtype:PCIe traddr:0000:04:00.0' -i 1 -c 4
Starting SPDK v19.01-pre / DPDK 18.08.0 initialization...
[ DPDK EAL parameters: perf -c 4 --base-virtaddr=0x200000000000 --file-prefix=spdk1 --proc-type=auto ]
EAL: Detected 40 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: Auto-detected process type: SECONDARY
EAL: Multi-process socket /run/user/1003/dpdk/spdk1/mp_socket_23851_100d594ccc17fe
EAL: Probing VFIO support...
Initializing NVMe Controllers
EAL: PCI device 0000:04:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
Attached to NVMe Controller at 0000:04:00.0 [8086:0953]
after io_queue_size: 256
controller IO queue size 256 less than required
Consider using lower queue depth or small IO size because IO requests may be queued at the NVMe driver.
WARNING: Some requested NVMe devices were skipped
Associating INTEL SSDPEDMD400G4  (PHFT620400TJ400BGN  ) with lcore 2
Initialization complete. Launching workers.
Starting thread on core 2
starting I/O failed
starting I/O failed
starting I/O failed
starting I/O failed
starting I/O failed
starting I/O failed
starting I/O failed
starting I/O failed
========================================================
                                                                                            Latency(us)
Device Information                                     :       IOPS       MB/s    Average        min        max
INTEL SSDPEDMD400G4  (PHFT620400TJ400BGN  ) from core 2:    2631.40    2631.40   21298.32   13186.93   25910.60
========================================================
Total                                                  :    2631.40    2631.40   21298.32   13186.93   25910.60

3. NVMe-oF configuration file:
NVMe-oF configuration file:
[Nvmf]
  AcceptorPollRate 10000

[Transport]
  Type RDMA

[Nvme]
  TransportID "trtype:PCIe traddr:0000:04:00.0" Nvme0

  RetryCount 4
  TimeoutUsec 0
  ActionOnTimeout None
  AdminPollRate 100000

  HotplugEnable No

[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Listen RDMA 10.10.16.9:4420
  AllowAnyHost Yes
  Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
  MaxNamespaces 20
  Namespace Nvme0n1 1

## Context (Environment including OS version, SPDK version, etc.)

OS: centos-release-7-3.1611.el7.centos.x86_64  kernel: 3.10.0-327.18.2.el7.x86_64  
spdk: commit ea002f5068f504b79a2f4a920655688d7bb15451

**Comment by jimharris** [comment]
Oh - I missed this from the output earlier in this thread:

"Consider using lower queue depth or small IO size because IO requests may be queued at the NVMe driver."

The -ENOMEM has nothing to do with hugepage memory nor primary/secondary processes.  The NVMe driver is saying that it won't be able to queue this level of I/O (64 I/O * 1MB).  The default queue size is 256 - but these SSDs have a max I/O size of 128KB.  Meaning each 1MB I/O results in 8 NVMe I/O.  64 * 8 = 512 which is bigger than the 256 queue depth size.  That's why this works at lower queue depth or smaller I/O size.

We could change perf to try allocating a bigger queue if needed.  These SSDs can allocate IO queues with more than 256 entries.

Could you confirm that you see this same issue just running perf as a primary process - without nvmf_tgt?

**Comment by ak-gh** [comment]
On Wed, Jan 9, 2019, 19:32 Jim Harris <notifications@github.com wrote:

> Oh - I missed this from the output earlier in this thread:
>
> "Consider using lower queue depth or small IO size because IO requests may
> be queued at the NVMe driver."
>
> The -ENOMEM has nothing to do with hugepage memory nor primary/secondary
> processes. The NVMe driver is saying that it won't be able to queue this
> level of I/O (64 I/O * 1MB). The default queue size is 256 - but these SSDs
> have a max I/O size of 128KB.
>
By the way, is it hard-coded or configurable on per SSD basis? We could
have had such config under nvme secrion for each drive, for example.

Thanks,
Andrey

**Comment by yuezhu1** [comment]
Many thanks for the input. I tried the command like "./perf -q 64 -o 1048576 -w randread -t 5 -r 'trtype:PCIe traddr:0000:04:00.0' -i 1 without setting up NVMe-oF. It works fine. 

The output is attached:./perf -q 64 -o 1048576 -w randread -t 5 -r 'trtype:PCIe traddr:0000:04:00.0' -i 1 
Starting SPDK v19.01-pre / DPDK 18.08.0 initialization...
[ DPDK EAL parameters: perf -c 0x1 --base-virtaddr=0x200000000000 --file-prefix=spdk1 --proc-type=auto ]
EAL: Detected 40 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: Auto-detected process type: PRIMARY
EAL: Multi-process socket /run/user/1003/dpdk/spdk1/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
Initializing NVMe Controllers
EAL: PCI device 0000:04:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
Attaching to NVMe Controller at 0000:04:00.0
Attached to NVMe Controller at 0000:04:00.0 [8086:0953]
after io_queue_size: 4096
Associating INTEL SSDPEDMD400G4  (PHFT620400TJ400BGN  ) with lcore 0
Initialization complete. Launching workers.
Starting thread on core 0
========================================================
                                                                                            Latency(us)
Device Information                                     :       IOPS       MB/s    Average        min        max
INTEL SSDPEDMD400G4  (PHFT620400TJ400BGN  ) from core 0:    3229.60    3229.60   19799.53     691.89   30225.07
========================================================
Total                                                  :    3229.60    3229.60   19799.53     691.89   30225.07

**Comment by jimharris** [comment]
Now I understand the problem.  This debug print gave the clue:

after io_queue_size: 4096

For the multi-process case, this same debug print shows 256.

By default, the SPDK NVMe driver creates the I/O queues with maximum size.  But the bdev NVMe driver overrides this to use just 256 instead.  Realistically, you don't need anything more than 256 total queue depth to reach maximum IOPs on an SSD, so the bdev NVMe driver just uses 256 by default to save on memory consumption.

So in the primary process case, perf gets the default SPDK NVMe driver setting, meaning for this SSD it gets 4096 queue size.  But for the secondary process case, the primary process has set the default to 256, so that's what the secondary process gets and you get your -ENOMEM failure on 64 QD and 1MB IO size.

I think the solution here is for nvme perf tool to not just assume that the default queue size will be the max queue size.  That's fine when perf is running as a primary process, but not when it's running as a secondary process.  We do have the ability to specify a different IO queue size when allocating the I/O queue pair.

**Comment by jimharris** [comment]
@ak-gh - this issue is specific to the nvme/perf tool (not bdevperf).  nvme/perf is built around just the NVMe driver, and does not use config files at all - only the command line.

**Comment by yuezhu1** [comment]
> Now I understand the problem. This debug print gave the clue:
> 
> after io_queue_size: 4096
> 
> For the multi-process case, this same debug print shows 256.
> 
> By default, the SPDK NVMe driver creates the I/O queues with maximum size. But the bdev NVMe driver overrides this to use just 256 instead. Realistically, you don't need anything more than 256 total queue depth to reach maximum IOPs on an SSD, so the bdev NVMe driver just uses 256 by default to save on memory consumption.
> 
> So in the primary process case, perf gets the default SPDK NVMe driver setting, meaning for this SSD it gets 4096 queue size. But for the secondary process case, the primary process has set the default to 256, so that's what the secondary process gets and you get your -ENOMEM failure on 64 QD and 1MB IO size.
> 
> I think the solution here is for nvme perf tool to not just assume that the default queue size will be the max queue size. That's fine when perf is running as a primary process, but not when it's running as a secondary process. We do have the ability to specify a different IO queue size when allocating the I/O queue pair.

For specifying the IO queue size, I wonder if you mean specify the io queue size through _opts_ to function spdk_nvme_ctrlr_alloc_io_qpair(), as the example shown below.

struct spdk_nvme_io_qpair_opts opts;
spdk_nvme_ctrlr_get_default_io_qpair_opts(ns_ctx->entry->u.nvme.ctrlr, &opts, sizeof(opts));
opts.io_queue_size = 4096;
pdk_nvme_ctrlr_alloc_io_qpair(ns_ctx->entry->u.nvme.ctrlr, &opts,
               sizeof(opts));

I tried the way I mentioned above by adding "opts.io_queue_size = 4096;" before https://github.com/spdk/spdk/blob/7fecba0c0239938fc88330688f2fb58f28cb685c/examples/nvme/perf/perf.c#L826
However, I still got the same error when running the perf as the secondary process. Therefore, I wonder if I did this wrong. 

Thank you again,
Yue

**Comment by changpe1** [comment]
> spdk_nvme_ctrlr_alloc_io_qpair

There are two parameters in the opts, io_queue_size and io_queue_requests, the io_queue_size is negotiated with the  hardware, for your case, it's configured in the NVMf target, so you can increase io_queue_requests here, although the real IO queue size is 256, but the requests sent from perf tool are queued at the NVMe driver layer.

**Comment by jimharris** [comment]
@changpe1 is right - can you try setting both io_queue_size = 4096 (like you have already done) as well as io_queue_requests?

**Comment by yuezhu1** [comment]
> @changpe1 is right - can you try setting both io_queue_size = 4096 (like you have already done) as well as io_queue_requests?

Yes, after I change both of them to 4096. The perf works fine. 

However, the reason I raised up the question is due to my own implementation. In my own case, the out of memory error still exists after change both io_queue_size to 4096 (even I don't start NVMe-oF). The available memory for spdk_env_opts.mem_size is 2048 MB. And I also allocated a 256 MB spdk buffer as the destination of submitted data requests. I wonder if you have any ideas in your mind that may cause the error. Any ideas would be greatly appreciated.

**Comment by jimharris** [comment]
Hi @yuezhu1,

Can you describe more about your own implementation and your own case?  There are many different ways an out-of-memory error could be returned to your application.

Keeping this issue open as a bug - we should automatically override io_queue_size and io_queue_requests when allocating the IO qpair.

-Jim

**Comment by yuezhu1** [comment]
Hi @jimharris,

Thanks for checking back with me. In my case, I have a multi-threads implementation. On every node, I also create an NVMe-oF Target. For my own implementation, I set the core mask for spdk_env_opts_init() on core 0 and 1, and the master core as 0. I also create a pthread and set its core affinity as 1 (called thread 1). I pre-allocated a 256 MB read buffer on hugepage through spdk_dma_zmalloc(). The submission qpair is created on main process through master core. Then, I submit read request on pthread 1, and the submission pattern follows the "perf" example (first submit N requests and poll completion, then the new request is issued in the call back function.)

I noticed a very interesting thing. When I was using INTEL SSDPEDMD400G4, the error appears no matter how I changed the io_queue_size. However, when I switched to INTEL SSDPED1D480GA, the problem disappears. I am not sure if this is related to the devices themselves. 

Thank you,
Yue
